x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/tokenizers/BasicTokenizer.java

```diff
@@ -0,0 +1,273 @@
+
+        int lastSplit = 0;
+        for (int i=0; i<codePoints.length; i++) {
+            if (isPunctuation(codePoints[i])) {
+                int charCount = i - lastSplit;
+                if (charCount > 0) {
+                    // add a new string for what has gone before
+                    split.add(new String(codePoints, lastSplit, i - lastSplit));
+                }
+                split.add(new String(codePoints, i, 1));
+                lastSplit = i+1;
+            }
+        }
+
+        if (lastSplit < codePoints.length) {
+            split.add(new String(codePoints, lastSplit, codePoints.length - lastSplit));
+        }
+
+        return split;
+    }
+
+    /**
+     * Surrounds any CJK character with whitespace
+     * @param text To tokenize
+     * @return tokenized text
+     */
+    static String tokenizeCjkChars(String text) {
+        StringBuilder sb = new StringBuilder(text.length());
+        boolean cjkCharFound = false;
+
+        int[] codePoints = text.codePoints().toArray();
+        for (int cp : codePoints) {
+            if (isCjkChar(cp)) {
+                sb.append(' ');
+                sb.appendCodePoint(cp);
+                sb.append(' ');
+                cjkCharFound = true;
+            } else {
+                sb.appendCodePoint(cp);
+            }
+        }
+
+        // no change
+        if (cjkCharFound == false) {
+            return text;
+        }
+
+        return sb.toString();
+    }
+
+    /**
+     * Remove control chars and normalize white space to ' '
+     * @param text Text to clean
+     * @return Cleaned text
+     */
+    static String cleanText(String text) {
+        int [] codePoints = text.codePoints()
+            .filter(codePoint -> (codePoint == 0x00 || codePoint == 0xFFFD || isControlChar(codePoint)) == false)
+            .map(codePoint -> isWhiteSpace(codePoint) ? ' ' : codePoint)
+            .toArray();
+
+        return new String(codePoints, 0, codePoints.length);
+    }
+
+    static boolean isCjkChar(int codePoint) {
+        // https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_(Unicode_block)
+        return (codePoint >= 0x3400 && codePoint <= 0x4DBF) || // CJK Unified Ideographs Extension A
+            (codePoint >= 0xF900 && codePoint <= 0xFAFF) ||   // CJK Compatibility Ideographs
+            (codePoint >= 0x4E00 && codePoint <= 0x9FFF) ||   // CJK unified Ideographs
+            // TODO extensions C, D, E look contiguous
+            (codePoint >= 0x20000 && codePoint <= 0x2A6DF) || // CJK Unified Ideographs Extension B
+            (codePoint >= 0x2A700 && codePoint <= 0x2B73F) || // CJK Unified Ideographs Extension C
+            (codePoint >= 0x2B740 && codePoint <= 0x2B81F) || // CJK Unified Ideographs Extension D
+            (codePoint >= 0x2B820 && codePoint <= 0x2CEAF) || // CJK Unified Ideographs Extension E
+            (codePoint >= 0x2F800 && codePoint <= 0x2FA1F);   // CJK Compatibility Ideographs Supplement
+    }
```

Comment:
```suggestion
    static boolean isCjkChar(int codePoint) {
        // https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_(Unicode_block)
        Character.UnicodeBlock block = Character.UnicodeBlock.of(codePoint);
        return Character.UnicodeBlock.CJK_COMPATIBILITY_IDEOGRAPHS.equals(block) ||
                Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS.equals(block) ||
                Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_A.equals(block) ||
                Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_B.equals(block) ||
                Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_C.equals(block) ||
                Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_D.equals(block) ||
                Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_E.equals(block) ||
                Character.UnicodeBlock.CJK_COMPATIBILITY_IDEOGRAPHS_SUPPLEMENT.equals(block);
    }
```

I don't think we should attempt to recreate the unicode blocks when they are standardized and we already have access to them.